packages/experimental/smart-session/src/core/controller/SmartSessionController.ts

```diff
@@ -0,0 +1,49 @@
+import { proxy, subscribe as sub } from 'valtio/vanilla'
+import type {
+  SmartSessionGrantPermissionsRequest,
+  SmartSessionGrantPermissionsResponse
+} from '../utils/TypeUtils.js'
+import { AppKitSmartSessionControllerClient } from '../../client.js'
+
+// -- Types --------------------------------------------- //
+export interface SmartSessionControllerState {
+  permissionsContext?: SmartSessionGrantPermissionsResponse['context']
+  permissions?: SmartSessionGrantPermissionsResponse['permissions']
+}
+
+export type StateKey = keyof SmartSessionControllerState
+
+// -- State --------------------------------------------------------- //
+const state = proxy<SmartSessionControllerState>({
+  permissions: undefined,
+  permissionsContext: undefined
+})
+
+// -- Controller ---------------------------------------- //
+export const SmartSessionController = {
+  state,
+
+  async grantPermissions(
```

Comment:
We don't validate the request in any way, with supported permissions etc

Reply:
We can also do this with the co-signer. That might be better, but then we need nice UX for the user

Reply:
Added validation of request and signer.